bitcoin/src/crypto/sighash.rs

```diff
@@ -1242,6 +1238,105 @@ impl<'a> Encodable for Annex<'a> {
     }
 }
 
+/// Error computing a taproot sighash.
+#[derive(Debug, Clone, PartialEq, Eq)]
+#[non_exhaustive]
+pub enum TaprootError {
+    /// Index out of bounds when accessing transaction input vector.
+    InputsIndex(transaction::InputsIndexError),
+    /// Can happen only when using `*_encode_signing_*` methods with custom writers, engines
+    /// like those used in `*_signature_hash` methods do not error.
+    Io(io::ErrorKind),
```

Comment:
Here's an idea, let's have:

```rust
pub enum Error<E> {
    Io(io::Error),
    Sighash(E),
}

impl<E> Error<E> {
    fn unwrap_sighash(self) -> E {
        match self {
            Self::Sighash(error) => error,
            Self::Io(error) => panic!("hash engine error {}", error),
        }
    }
}
```

Then we can retain the derives on this type even when it's `io::Error` and call `unwrap_sighash` in methods returning hashes. It also avoids duplication of the `Io` variant across all enums.

Reply:
At first blush, that looks like a good idea.